---
title: "Quantitative cross-linking data processing"
output:
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---



```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(readxl)
library(ggplot2) 
library(plotly)
library(ggpubr)
library(ggforce)
library(extrafont)
library(rawR)
library(ggrepel)

```



Results: Details of files to analyse and run information
Note: "FWD" means that the heavy labelling was performed on the wild-type

```{r warning=FALSE}
print(paste0("Report generated on:  ", Sys.Date()))

RawXLinkProphetOutput <- read.delim("Data/XLinkProphet/iprophet-xl.xls", stringsAsFactors = FALSE)

#generated by running convert convertXLinkProphetXLStoUpload6col.pl iprophet-xl.xls 0.3065 FASTA=C:\TPP\database\yeastproteome.fasta
Raw6colOutput <- read.delim("Data/XLinkProphet/iprophet-xl_6col.txt", header=FALSE, stringsAsFactors = FALSE)
colnames(Raw6colOutput) <- c("Pep1", "Accession1", "PepPos1", "Pep2", "Accession2", "PepPos2", "ProtPos1", "ProtPos2")

SampleName = "04112020_Hpm1KO_Spheroplast"

#probability from XLinkProphet for CSMs to get a 1% FDR. This is different in every sample.
probability_01 = 0.0883
probability_comp_01 = 0.3065

#MethylQuant confidence level
MethylQuantConfidence = "High"
```



Filtering the peptide prophet results of interest by the probablity as assigned by XLinkProphet
```{r}
FilteredCSMs <- RawXLinkProphetOutput %>% 
  filter(probability >= probability_01)

FilteredRedundantCrosslinks <- RawXLinkProphetOutput %>% 
  filter(composite_probability >= probability_comp_01) %>% 
  distinct()
```

Processing to combined IDs and generate different levels of redundancy
```{r}
ReturnXL_Stripped <- function(PeptideWithMods) {
  StrippedPep <- str_remove_all(PeptideWithMods, "\\[(.*?)\\]")
       return(StrippedPep)
}

AlphabetizeXLKey <- function(XLPep1, XLPep2) {
  Interaction <- if_else(XLPep1 == XLPep2, paste0(XLPep1, "_", XLPep2), if_else(XLPep1 < XLPep2, paste0(XLPep1, "_", XLPep2), paste0(XLPep2, "_", XLPep1)))
	return(Interaction)
}

AddKey <- function( input_table, Level) {
  nrows <- nrow(input_table)
    if(Level == "XL") 
  {
    for (i in 1:nrows) {
	  input_table$Pep1[i] <- ReturnXL_Stripped(input_table$Pep1[i])
	  input_table$Pep2[i] <- ReturnXL_Stripped(input_table$Pep2[i])
	  input_table$XLPep1[i] <- paste0(input_table$Pep1[i], "-", input_table$PepPos1[i], "-", input_table$Accession1[i])
	  input_table$XLPep2[i] <- paste0(input_table$Pep2[i], "-", input_table$PepPos2[i], "-", input_table$Accession2[i])
	  input_table$XL_Key[i] <- AlphabetizeXLKey(input_table$XLPep1[i], input_table$XLPep2[i])
	}
	return(input_table)
  }
  
  else if (Level == "URP") {
    for (i in 1:nrows) {
	  input_table$Pep1[i] <- ReturnXL_Stripped(input_table$Pep1[i])
	  input_table$Pep2[i] <- ReturnXL_Stripped(input_table$Pep2[i])
	  input_table$ResidueKey1[i] <- paste0(input_table$Accession1[i], "-", input_table$ProtPos1[i])
	  input_table$ResidueKey2[i] <- paste0(input_table$Accession2[i], "-", input_table$ProtPos2[i])
	  input_table$URP_Key[i] <- AlphabetizeXLKey(input_table$ResidueKey1[i], input_table$ResidueKey2[i])
    }
    return(input_table)
  }
  
  else if (Level == "PPI") {
    for (i in 1:nrows) {
	  input_table$PPI_Key[i] <- AlphabetizeXLKey(input_table$Accession1[i], input_table$Accession2[i])
    }
    return(input_table)
  }
}

#Useful functions to return the 0-based position of the crosslink mod (325.13 = light PIR stump mass, 333.14 = heavy labelled stump mass)
#or strip the peptide sequence, as exported by XLinkProphet under the heading peptide1/peptide2, of any mods. Useful for XLinkDB upload.
ReturnXL_PepPos <- function(PeptideWithMods) {
    V1 <- str_replace(PeptideWithMods, "\\[325\\.13\\]", "x")
    V1 <- str_replace(V1, "\\[333\\.14\\]", "x")
    V1 <- str_remove_all(V1, "\\[(.*?)\\]")
    V1 <- str_locate(V1, "x")[1]

    #make 0-based modPos
    V1 <- V1 -2
    return(V1);
}

CleanXLinkProphet <- function(XlinkProphet_output) {
  nrows <- nrow(XlinkProphet_output)

  for (i in 1:nrows) {
  XlinkProphet_output$Accession1[i] <- str_remove(str_sub(XlinkProphet_output$protein1[i], 1, 9), "sp\\|" ) 
  XlinkProphet_output$Accession2[i] <-  str_remove(str_sub(XlinkProphet_output$protein2[i], 1, 9), "sp\\|" ) 
  XlinkProphet_output$PepPos1[i] <- ReturnXL_PepPos(XlinkProphet_output$peptide1[i])
  XlinkProphet_output$PepPos2[i] <- ReturnXL_PepPos(XlinkProphet_output$peptide2[i])
  XlinkProphet_output$Pep1[i] <- ReturnXL_Stripped(XlinkProphet_output$peptide1[i])
  XlinkProphet_output$Pep2[i] <- ReturnXL_Stripped(XlinkProphet_output$peptide2[i])
    }

  return(XlinkProphet_output);
}

CleanXlinkProphetTable <- AddKey(CleanXLinkProphet(FilteredCSMs), "XL")
FilteredRedundantCrosslinks <- AddKey(CleanXLinkProphet(FilteredRedundantCrosslinks), "XL")



Clean6colTable <- AddKey(Raw6colOutput, "PPI")
Clean6colTable <- AddKey(Clean6colTable, "URP")
Clean6colTable <- AddKey(Clean6colTable, "XL")
Clean6colTable <- Clean6colTable %>%  select(XL_Key, URP_Key, PPI_Key, ResidueKey1, ResidueKey2)

AnnotatedXLTable <- distinct(full_join(CleanXlinkProphetTable, Clean6colTable, by = "XL_Key" ))
write.table(AnnotatedXLTable, file = paste0(SampleName, "_AnnotatedXLTable.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
```

```{r}
APID <- read.delim("Data/APID/559292_noISI_Q3_June2021.txt", stringsAsFactors=FALSE)

APID <- APID %>% 
  filter(ExpEvidences >= 1) %>% 
  rename(Accession1 = UniprotID_A, Accession2 = UniprotID_B) %>% 
  mutate(KnownAPID = TRUE)

APID <- AddKey(APID, "PPI")


PPIs <- AnnotatedXLTable %>% 
  filter(intra != 1) %>% 
  filter(!is.na(PPI_Key)) %>% 
  group_by(PPI_Key) %>% 
  summarise(NumURPs = n_distinct(URP_Key), NumCSMs = n()) %>% 
  distinct()

PPI_URPs <- AnnotatedXLTable %>% 
  filter(intra != 1) %>% 
    filter(!is.na(PPI_Key)) %>% 
  select(PPI_Key, URP_Key) %>% 
  distinct()

PPIs <- left_join(PPIs, APID, by = "PPI_Key")

PPIs <- PPIs %>% 
  mutate(KnownAPID = replace_na(KnownAPID, FALSE)) %>% 
  separate(PPI_Key, into = c("AccessionA", "AccessionB"), sep = "_",remove= FALSE )

PPI_URPs <- left_join(PPI_URPs, PPIs, by = "PPI_Key")

write.table(PPIs, file = "IdentifiedPPIs.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(PPI_URPs, file = "IdentifiedPPI_URPs.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

Generating a table of URPs.
```{r}
URPs <- AnnotatedXLTable %>% 
  filter(composite_probability >= probability_comp_01) %>% 
  mutate(CrosslinkType = if_else(intra == 1, "Intra", "Inter")) %>% 
  select(URP_Key, CrosslinkType) %>% 
  separate(URP_Key, sep = "_", into = c("Residue1", "Residue2"), remove = FALSE) %>% 
  separate(Residue1, sep = "-", into = c("Accession1", "Lys1"), remove = FALSE) %>% 
  separate(Residue2, sep = "-", into = c("Accession2", "Lys2"), remove = FALSE) %>% 
  distinct()  


write.table(URPs, file = "URPs.txt", row.names = FALSE, quote = FALSE, sep = "\t")
```


Make an input table for XLinkDB
```{r}
XLinkDBTable <- AnnotatedXLTable %>% 
  select(Pep1, Accession1, PepPos1, Pep2, Accession2, PepPos2) %>% 
  distinct()

write.table(XLinkDBTable, file = paste0("XLinkDBInput_", SampleName, ".txt"), sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE)  
```



Generating a table of CSMs compatible for analysis with MethylQuant
Converting the calculated crosslink monoisotopic parent mass (theoretically determined by Comet when using C13 isotope error in search) to a m/z and calculate the theoretical isotope labelling pair. This is useful for use with MethylQuant, where you provide a list of m/z, charges, scans and mass differences (actually, a m/z difference calculated by taking into account the number of arginines and lysines present in the sequence, and the parent charge).
Searching for m/z's of real spectral identifications, and also for masses in other replicates (matches between runs idea).

```{r}
TableForMethylQuant <- AnnotatedXLTable %>% 
  mutate(isLIGHT_ID = str_detect(peptide1, "325.13")) %>% 
  mutate(Sequence = str_replace_all(composite_id, "_", "")) %>% 
  mutate(Sequence = str_replace_all(Sequence, "x", "")) %>% 
  mutate(NumLysines = str_count(Sequence, "K")) %>% 
  mutate(NumArginines = str_count(Sequence, "R")) %>% 
  mutate(mono_parent_mass = peptide1_mass + peptide2_mass + 693.399601) %>% 
  mutate(HeavyMZshift = ((NumLysines * 8.014199) + (NumArginines * 10.008269))/parent_charge) %>% 
  mutate(MZ = (mono_parent_mass + (parent_charge * 1.00728))/parent_charge) %>% 
  mutate(MZShift = ifelse(isLIGHT_ID == FALSE, HeavyMZshift * -1, HeavyMZshift)) %>% 
  mutate(Modifications = "") %>%
  rename(Charge = `parent_charge`) %>% 
  rename("Calc m/z" = MZ) %>% 
  rename("Mass Difference" = MZShift) %>% 
  mutate(realID = TRUE)

nrows <- nrow(TableForMethylQuant)
for (i in 1:nrows) {
  TableForMethylQuant$SpecID[i] <- tail(strsplit(TableForMethylQuant$spectrum[i],"\\\\")[[1]],1)
  TableForMethylQuant$'Data File'[i] <- paste0(str_split_fixed(TableForMethylQuant$SpecID[i], "\\.", n =4)[,1], ".raw")
  TableForMethylQuant$'Start Scan'[i] <- str_split_fixed(TableForMethylQuant$SpecID[i], "\\.", n =4)[,2]
} 

TableForMethylQuant <- TableForMethylQuant %>% 
  filter(!is.na(`Data File`))

DuplicateRV_1 <- TableForMethylQuant %>% 
  filter(!(str_detect(`Data File`, "RV_") & str_detect(`Data File`, "_1.raw"))) %>% 
  mutate(`Data File` = str_replace(`Data File`, "FWD", "RV")) %>% 
  mutate(`Data File` = str_replace(`Data File`, "_2.raw", "_1.raw")) %>% 
  select(Accession1:`Start Scan`, peptide1_mass, peptide2_mass, parent_mass, Charge) %>% 
  select(-SpecID) %>% 
  mutate(realID = FALSE) 

DuplicateRV_2 <- TableForMethylQuant %>% 
  filter(!(str_detect(`Data File`, "RV_") & str_detect(`Data File`, "_2.raw"))) %>% 
  mutate(`Data File` = str_replace(`Data File`, "FWD", "RV")) %>% 
  mutate(`Data File` = str_replace(`Data File`, "_1.raw", "_2.raw")) %>% 
  select(Accession1:`Start Scan`, peptide1_mass, peptide2_mass, parent_mass, Charge) %>% 
  select(-SpecID) %>% 
  mutate(realID = FALSE) 


DuplicateFWD_1 <- TableForMethylQuant %>% 
  filter(!(str_detect(`Data File`, "FWD_") & str_detect(`Data File`, "_1.raw"))) %>% 
  mutate(`Data File` = str_replace(`Data File`, "RV", "FWD")) %>% 
  mutate(`Data File` = str_replace(`Data File`, "_2.raw", "_1.raw")) %>% 
  select(Accession1:`Start Scan`, peptide1_mass, peptide2_mass, parent_mass, Charge) %>% 
  select(-SpecID) %>% 
  mutate(realID = FALSE) 


DuplicateFWD_2 <- TableForMethylQuant %>% 
  filter(!(str_detect(`Data File`, "FWD_") & str_detect(`Data File`, "_2."))) %>% 
  mutate(`Data File` = str_replace(`Data File`, "RV", "FWD")) %>% 
  mutate(`Data File` = str_replace(`Data File`, "_1.raw", "_2.raw")) %>% 
  select(Accession1:`Start Scan`, peptide1_mass, peptide2_mass, parent_mass, Charge) %>% 
  select(-SpecID) %>% 
  mutate(realID = FALSE) 

TableForMethylQuant <- bind_rows(TableForMethylQuant, DuplicateRV_1, DuplicateRV_2, DuplicateFWD_1, DuplicateFWD_2)

TableForMethylQuant <- TableForMethylQuant %>% 
    select(Accession1:`Start Scan`, peptide1_mass, peptide2_mass, parent_mass, Charge) %>% 
  distinct() %>% 
  filter(!is.na(`Data File`)) %>% 
  filter(!is.na(Sequence))

TableForMethylQuant <- distinct(TableForMethylQuant)

write.csv(TableForMethylQuant, file = paste0("TableForMethylQuant_IDsDuplicated", SampleName, ".csv"), row.names = FALSE, quote = TRUE)  
```

Protein level normalisation factor extrapolation.
Taking the normalisation factor for crosslink ratios by extrapolating normalisation factor from MaxQuant results of the matched proteome analysis.
```{r}
proteinGroupsFWD <- read.delim("Data/MaxQuant/FW/proteinGroups.txt", stringsAsFactors=FALSE)
proteinGroupsRV <- read.delim("Data/MaxQuant/RV/proteinGroups.txt", stringsAsFactors=FALSE)

proteinGroupsFWD <- proteinGroupsFWD %>% 
  mutate(Replicate = "FWD")  

proteinGroupsRV <- proteinGroupsRV %>% 
  mutate(Replicate = "RV")  

ProteinswithRatiosFWD <- proteinGroupsFWD %>% 
  filter(Ratio.H.L<100)

FWD.meanProteinH.L = mean(ProteinswithRatiosFWD$Ratio.H.L)
FWD.meanProteinH.L

ProteinswithRatiosRV <- proteinGroupsRV %>% 
  filter(Ratio.H.L<100)

RV.meanProteinH.L = mean(ProteinswithRatiosRV$Ratio.H.L)
RV.meanProteinH.L

```


MethylQuant output cleaning, to check the quantified species match input by comparing PPM, and also filtering quanification above a score of 10.
```{r}
MethylQuantOutput <- read_csv("Data/MethylQuant/TableForMethylQuant_IDsDuplicated04112020_Hpm1KO_Spheroplast_MethylQuant.csv")

MethylQuantOutputLight <- MethylQuantOutput %>% 
  filter(isLIGHT_ID == TRUE) %>% 
  mutate(LightMZ =`Calc m/z`) %>% 
  mutate(MethylQuantMassDiffLight = ((`Light m/z 1 IsotopeCorrelation`- LightMZ))) %>%  
  mutate(MethylQuantMassDiffLightPPM = abs(((`Light m/z 1 IsotopeCorrelation`- LightMZ)/LightMZ)*1000000) ) %>% 
  mutate(HeavyMZ = `Calc m/z` + `Mass Difference`) %>% 
  mutate(MethylQuantMassDiffHeavy = ((`Heavy m/z 1 IsotopeCorrelation`- HeavyMZ))) %>%  
  mutate(MethylQuantMassDiffHeavyPPM = abs(((`Heavy m/z 1 IsotopeCorrelation`-  HeavyMZ)/HeavyMZ)*1000000) ) 

MethylQuantOutputHeavy <- MethylQuantOutput %>% 
  filter(isLIGHT_ID == FALSE) %>%
  mutate(LightMZ = `Calc m/z` + `Mass Difference`) %>% 
  mutate(MethylQuantMassDiffLight = ((`Light m/z 1 IsotopeCorrelation`- LightMZ))) %>%  
  mutate(MethylQuantMassDiffLightPPM = abs(((`Light m/z 1 IsotopeCorrelation`- LightMZ)/LightMZ)*1000000) ) %>% 
  mutate(HeavyMZ =  `Calc m/z`) %>% 
  mutate(MethylQuantMassDiffHeavy = ((`Heavy m/z 1 IsotopeCorrelation`- HeavyMZ))) %>%  
  mutate(MethylQuantMassDiffHeavyPPM = abs(((`Heavy m/z 1 IsotopeCorrelation`-  HeavyMZ)/HeavyMZ)*1000000) ) 

MethylQuantOutput <- bind_rows(MethylQuantOutputLight, MethylQuantOutputHeavy)

MethylQuantOutput <- MethylQuantOutput %>% 
  mutate(Iso2Light = LightMZ + (1.00866491588/Charge)) %>% 
  mutate(Iso3Light = Iso2Light + (1.00866491588/Charge)) %>% 
  mutate(Iso2LightPPM = ((Iso2Light -  `Light m/z 2 IsotopeCorrelation`)/Iso2Light)*100000) %>% 
  mutate(Iso3LightPPM = ((Iso3Light -  `Light m/z 3 IsotopeCorrelation`)/Iso3Light)*100000) %>% 
  mutate(Iso2Heavy = HeavyMZ + (1.00866491588/Charge)) %>% 
  mutate(Iso3Heavy = Iso2Heavy + (1.00866491588/Charge)) %>% 
  mutate(Iso2HeavyPPM = ((Iso2Heavy -  `Heavy m/z 2 IsotopeCorrelation`)/Iso2Heavy)*100000) %>% 
  mutate(Iso3HeavyPPM = ((Iso3Heavy-  `Heavy m/z 3 IsotopeCorrelation`)/Iso3Heavy)*100000) 


MethylQuantOutput <- MethylQuantOutput %>% 
  mutate(Replicate = if_else(str_detect(`Data File`, "RV_"), "RV", "FWD")) %>% 
  mutate(XL_ID = row_number())

PepsWithRatios <- MethylQuantOutput %>% 
  filter(!is.na(`H/L Ratio #2`)) %>% 
  filter(MethylQuantMassDiffHeavyPPM <= 10) %>% 
  filter(MethylQuantMassDiffLightPPM <= 10) %>% 
  filter(`MethylQuant Score` > 10)

```

Remove crosslinked peptides containing methionine oxidation, which may interfere with quantification

```{r}
MethioninePeps <- AnnotatedXLTable %>% 
  filter(str_detect(composite_id, "147.04")) 

  for (i in 1:NROW(MethioninePeps)) {
     MethioninePeps$spectrum[i] = tail(strsplit(MethioninePeps$spectrum[i],"\\\\")[[1]],1)
}

PepsWithRatios <- anti_join(PepsWithRatios, MethioninePeps, by = c("SpecID" = "spectrum"))
```

```{r}
FilteredRedundantCrosslinks <- FilteredRedundantCrosslinks %>% 
  select(XL_Key) %>% 
  mutate(ValidCrosslink = TRUE) %>% 
  distinct()

PepsWithRatios <- inner_join(PepsWithRatios, FilteredRedundantCrosslinks, by = "XL_Key" )

PepsWithRatios <- distinct(PepsWithRatios)
```

Normalising SILAC ratios using protein level normalisation factor, which was was the mean protein H/L ratio from matched proteomics experiments.
Summarising to URPs and only taking URP that were quantified in both replicates.
```{r warning=FALSE}
URPsQuantifiedInBoth <- PepsWithRatios %>% 
  select(XL_Key, URP_Key, PPI_Key,  `Data File`, Replicate, `H/L Ratio #2`, `MethylQuant Score`) %>% 
  group_by(XL_Key, URP_Key, PPI_Key,  `Data File`, Replicate) %>% 
  top_n(n = 1, wt = `MethylQuant Score`) %>%
  distinct() %>% 
  ungroup() %>% 
  mutate(NormCrosslinkRatio = if_else(Replicate == "RV", `H/L Ratio #2`/RV.meanProteinH.L, `H/L Ratio #2`/FWD.meanProteinH.L)) %>% 
  mutate(NormCrosslinkLog2Ratio = log2(NormCrosslinkRatio)) %>%
  mutate(NormCrosslinkLog2_KO_WT_Ratio = if_else(Replicate == "RV", NormCrosslinkLog2Ratio * -1, NormCrosslinkLog2Ratio )) %>% 
  select(XL_Key, URP_Key, PPI_Key, NormCrosslinkLog2_KO_WT_Ratio,  `Data File`,  Replicate) %>% 
  group_by(URP_Key, PPI_Key) %>% 
  summarise(AvgCorrectXLRatio = mean(NormCrosslinkLog2_KO_WT_Ratio), 
            MedianCorrectXLRatio = median(NormCrosslinkLog2_KO_WT_Ratio), 
            NumRatios = n(), 
            sd = sd(NormCrosslinkLog2_KO_WT_Ratio), 
            Ratios = list(NormCrosslinkLog2_KO_WT_Ratio), 
            RatiosWritten = paste(Ratios, collapse=', ' ),
            NumBioReplicates = n_distinct(Replicate)) %>% 
  ungroup() %>% 
  separate(URP_Key, c("ResidueKey1", "ResidueKey2"), sep = "_", remove = FALSE)  %>% 
  separate(ResidueKey1, c("Accession1", "Residue1"), sep = "-", remove = FALSE) %>% 
  separate(ResidueKey2, c("Accession2", "Residue2"), sep = "-", remove = FALSE) %>% 
  filter(NumRatios > 1) %>% 
  filter(NumBioReplicates == 2)
```

Performing two-tailed, unpaired t-tests to compare crosslink ratios to a ratio value of 0 (no change).
Using the Benjamini-Hochberg ("fdr") method correct for multiple testing, generating adjusted p-values.
```{r}
URPsQuantifiedInBoth$Pval = NA

for (i in 1:NROW(URPsQuantifiedInBoth)){

  Ratios = unlist(URPsQuantifiedInBoth$Ratios[i])
  TTestResult = t.test(Ratios, mu = 0, alternative = "two.sided")
  URPsQuantifiedInBoth$Pval[i] = TTestResult$p.value[1]

}

AdjustedPValues <- data.frame()
AdjustedPValues <- p.adjust(URPsQuantifiedInBoth$Pval, method = "fdr")

URPsQuantifiedInBoth$Adj.Pval <- AdjustedPValues
```

Annotating information about proteins from UNIPROT
```{r}
UNIPROT <- read.delim("Data/UniProt/UNIPROT_Yeast.txt", stringsAsFactors=FALSE)
UNIPROT <- UNIPROT %>% 
  mutate(Gene.names.primary.1 = paste0(str_to_sentence(Gene.names...primary.., locale = "en"), "p")) %>% 
  rename(Entry.name.1 = Entry.name) %>% 
  select(Entry, Entry.name.1, Gene.names.primary.1) 


URPsQuantifiedInBoth <- left_join(URPsQuantifiedInBoth, UNIPROT, by = c("Accession1" = "Entry"))

UNIPROT <- UNIPROT %>% 
  rename(Entry.name.2 = Entry.name.1,
         Gene.names.primary.2 = Gene.names.primary.1)
URPsQuantifiedInBoth <- left_join(URPsQuantifiedInBoth, UNIPROT, by = c("Accession2" = "Entry"))

URPsQuantifiedInBothTable <- URPsQuantifiedInBoth %>% 
  select(-Ratios) %>% 
  rename(Ratios = RatiosWritten)

write.table(URPsQuantifiedInBothTable, file = "URPsQuantifiedInBoth.txt", sep = "\t", row.names = FALSE, quote = FALSE)

```

Generating volcano plot.
Purposely excluding one extreme value for visibility sake
```{r}
ForVolcano <- URPsQuantifiedInBoth %>% 
    mutate(log10pvalue = -1* log10(Adj.Pval)) %>% 
    mutate(Direction = case_when(
    (Adj.Pval <= 0.05) & (AvgCorrectXLRatio > 0) ~ "Up",
    (Adj.Pval <= 0.05) & (AvgCorrectXLRatio < 0) ~ "Down",
    (Adj.Pval > 0.05)  ~"Not")) %>% 
    mutate(LabelName = paste0(Gene.names.primary.1, "-\n", Gene.names.primary.2)) %>% 
    #mutate(LabelName =  Gene.names.primary.1) %>% 
  mutate(Label = if_else(((Adj.Pval < 0.05) & (abs(AvgCorrectXLRatio >= 0.7))), TRUE, FALSE)) %>% 
    mutate(Alpha = if_else(Adj.Pval <= 0.05, TRUE, FALSE))



ggplot(ForVolcano, aes(x = AvgCorrectXLRatio, y = log10pvalue,  color = Direction ,alpha = Alpha, label = LabelName)) +
  geom_jitter() +
  #geom_text_repel(data = subset(ForVolcano, Label == TRUE), size = 3) +
  scale_color_manual(values = c( "#0000FF", "#bababa", "#D90000")) +
  theme_bw() +
  theme(legend.position = "none", 
        text=element_text(size=10,  family="Arial"), 
        panel.border = element_rect(fill=NA, colour = "black", size=1),
        panel.grid.minor = element_blank()) +
  #xlim(-2.3,2.3)+ 
 ylim(0, 3.5)
  
ggsave(filename = "CrosslinkVolcano.pdf", device = cairo_pdf, 
       width = 3, height = 3, units = "in")
```



Grouping URPs by signifcance level
```{r}
SignificantlyChanging_URPs <- URPsQuantifiedInBoth %>% 
  filter(Adj.Pval <= 0.05) 

SignificantlyChanging_URPs

NonChanging_URPS <- URPsQuantifiedInBoth %>% 
  filter(Adj.Pval > 0.05) 
NonChanging_URPS
```

Annoting protein level change to changing URPs
```{r}
proteinGroupsCombined <- bind_rows(proteinGroupsFWD, proteinGroupsRV)

filteredProteinGroups <- proteinGroupsCombined %>% 
  separate_rows(Majority.protein.IDs, sep =";") %>% 
  filter(!is.na(Ratio.H.L.normalized)) %>% 
  filter(Ratio.H.L.variability.... < 30) %>% 
  filter(!str_detect(Protein.IDs, "CON_")) %>% 
  filter(!str_detect(Protein.IDs, "REV_")) %>% 
  mutate(Accession1 =  str_remove(str_sub(Majority.protein.IDs, 1, 9), "sp\\|" )) %>% 
  rename(NormProteinRatio = Ratio.H.L.normalized) %>% 
  mutate(NormProteinLog2Ratio = log2(NormProteinRatio)) %>% 
  mutate(NormProteinLog2_KO_WT_ratio = if_else(Replicate == "RV", NormProteinLog2Ratio * -1, NormProteinLog2Ratio)) %>% 
  select(NormProteinLog2_KO_WT_ratio, Accession1, Fasta.headers, Ratio.H.L, Ratio.H.L.variability...., Ratio.H.L.count, Replicate)

#average over replicates
MeanRatioProteinLevel <- filteredProteinGroups %>% 
  group_by(Accession1) %>% 
  summarise(Mean_NormProteinLog2_KO_WT_ratio.Protein.1 = mean(NormProteinLog2_KO_WT_ratio))

SignificantlyChanging_URPs <- left_join(SignificantlyChanging_URPs, MeanRatioProteinLevel, by = c("Accession1" = "Accession1"))

MeanRatioProteinLevel <- MeanRatioProteinLevel %>% 
  rename(Mean_NormProteinLog2_KO_WT_ratio.Protein.2 = Mean_NormProteinLog2_KO_WT_ratio.Protein.1)
SignificantlyChanging_URPs <- left_join(SignificantlyChanging_URPs, MeanRatioProteinLevel, by = c("Accession2" = "Accession1"))

```


Comparing URP ratios to protein level changes, two-tailed, unpaired t-test, with "fdr" method of multiple testing comparison
```{r}

SignificantlyChanging_URPs_withProteinQuant <- SignificantlyChanging_URPs %>% 
  filter(!is.na(Mean_NormProteinLog2_KO_WT_ratio.Protein.1)) %>% 
  filter(!is.na(Mean_NormProteinLog2_KO_WT_ratio.Protein.2)) 


for (i in 1:NROW(SignificantlyChanging_URPs_withProteinQuant)){

  Ratios = unlist(SignificantlyChanging_URPs_withProteinQuant$Ratios[i])
  
  Test.1 = t.test(Ratios, mu = SignificantlyChanging_URPs_withProteinQuant$Mean_NormProteinLog2_KO_WT_ratio.Protein.1[i], alternative = "two.sided")
  SignificantlyChanging_URPs_withProteinQuant$Pval.againstProtein.1[i] = Test.1$p.value[1]
  
  Test.2 = t.test(Ratios, mu = SignificantlyChanging_URPs_withProteinQuant$Mean_NormProteinLog2_KO_WT_ratio.Protein.2[i], alternative = "two.sided")
  SignificantlyChanging_URPs_withProteinQuant$Pval.againstProtein.2[i] = Test.2$p.value[1]
}

Adjusted1 <- data.frame()
Adjusted2 <- data.frame()

Adjusted1 <- p.adjust(SignificantlyChanging_URPs_withProteinQuant$Pval.againstProtein.1, method = "fdr")
Adjusted2 <- p.adjust(SignificantlyChanging_URPs_withProteinQuant$Pval.againstProtein.2, method = "fdr")

SignificantlyChanging_URPs_withProteinQuant$Adj.Pval.againstProtein.1 <- Adjusted1
SignificantlyChanging_URPs_withProteinQuant$Adj.Pval.againstProtein.2 <- Adjusted2
```

Generating a list of significantly changing URPs (at crosslink, and compared to protein level) for figure making
```{r}
ForFig <- SignificantlyChanging_URPs_withProteinQuant %>% 
  filter(Adj.Pval.againstProtein.1 <= 0.05) %>% 
  filter(Adj.Pval.againstProtein.2 <= 0.05) %>% 
  select(-Ratios) %>% 
  mutate(ProteinRatioToPlot = if_else(abs(Adj.Pval.againstProtein.1) > abs(Adj.Pval.againstProtein.2), Adj.Pval.againstProtein.1, Adj.Pval.againstProtein.2))

ForFig

write.table(ForFig, file = "CrosslinksForCytoscape.txt", sep = "\t", row.names = FALSE, quote = FALSE)

```


Investigating URPs with signal in only one strain, but also replicatable across label swap replicates
```{r}
PepsWithoutRatios <- anti_join(MethylQuantOutput, PepsWithRatios, by = "URP_Key")

PepsWithOnlyOneChanelOnly <- PepsWithoutRatios %>% 
  separate(ResidueKey1, c("Accession1", "Residue1"), sep = "-", remove = FALSE) %>% 
  separate(ResidueKey2, c("Accession2", "Residue2"), sep = "-", remove = FALSE) %>% 
  distinct() %>% 
  mutate(TotalLightIntensity = `Light Intensity 1 IsotopeCorrelation` + `Light Intensity 2 IsotopeCorrelation` + `Light Intensity 3 IsotopeCorrelation`) %>% 
  mutate(TotalHeavyIntensity = `Heavy Intensity 1 IsotopeCorrelation` + `Heavy Intensity 2 IsotopeCorrelation` + `Heavy Intensity 3 IsotopeCorrelation` ) %>% 
  mutate(OnlyHeavy = if_else(TotalLightIntensity == 0 & TotalHeavyIntensity !=0, TRUE, FALSE)) %>% 
  mutate(OnlyLight = if_else(TotalHeavyIntensity == 0 & TotalLightIntensity !=0, TRUE, FALSE)) %>% 
  group_by(URP_Key, Replicate, Accession1, Residue1, Accession2, Residue2) %>% 
  distinct() %>%
  summarise(AllOnlyLight = all(OnlyLight),
            AllOnlyHeavy = all(OnlyHeavy)) %>% 
  ungroup() %>% 
  mutate(UniqueToOneChannel = AllOnlyLight + AllOnlyHeavy) %>% 
  filter(UniqueToOneChannel == 1) %>% 
  group_by(URP_Key, Accession1, Residue1, Accession2, Residue2) %>%  
  summarise(NumberReps = n(), SumAllLight = sum(AllOnlyLight), SumAllHeavy = sum(AllOnlyHeavy)) %>% 
  ungroup() 

PepsWithOnlyChanelOnly_OppositeInLabelSwap <- PepsWithOnlyOneChanelOnly %>% 
  filter(NumberReps == 2, SumAllLight == 1, SumAllHeavy == 1)

NROW(PepsWithOnlyChanelOnly_OppositeInLabelSwap)
```

